### PR TITLE
ci: Integrate UATs to full-bundle-tests.yaml

### LINKED ISSUE #683: Run the Kubeflow UATs for `latest/edge` in airgapped
As part of https://github.com/canonical/bundle-kubeflow/issues/681 we have to verify that the `latest/edge` version of the bundle works in an airgapped deployment. Before we can do that, though, we need to check whether the UATs can run successfully on `latest/edge` (currently only tested with 1.7).

### Goal

Run the Kubeflow UATs on a `latest/edge` deployment to see what works and what doesn't.

### PR DESCRIPTION
**Pull request by orfeas-k**
This PR integrates the UATs to [full-bundle-tests.yaml](https://github.com/canonical/bundle-kubeflow/blob/main/.github/workflows/full-bundle-tests.yaml) workflow. The PR should be merged alongside https://github.com/canonical/charmed-kubeflow-uats/pull/56. Note that at the moment, the training-integration UAT will fail on self-hosted runners due to https://github.com/canonical/charmed-kubeflow-uats/issues/50. 
  
- Update [tests-bundle/1.8](https://github.com/canonical/bundle-kubeflow/tree/main/tests-bundle/1.8)
- Unpin juju in tests-bundle/1.8.
- Refactor workflow to use Aproxy, Python 3.8 and operator action
- Add UATs to workflow
- Add save debug artifacts to workflow

Fixes canonical/charmed-kubeflow-uats#7

### Testing the action
In order to see the action live, you could go to Actions > Deploy bundle and run UATs on self-hosted runners on this repo and trigger it with the following settings:
- branch: kf-4343-integrate-uats-to-kubeflow-bundle-ci
- test folder: tests-bundle/1.8
- channel: --channel=1.8/stable
- uats branch: kf-4343-trustenv-false-lightkube
- microk8s channel: 1.25-strict/stable (or w/e you wish) 
- Juju channel: 3.1/stable

### CODE REVIEW

File: .github/workflows/full-bundle-tests.yaml [hunk 1]
```diff
@@ -96,25 +92,40 @@ jobs:
...
+
+      - name: Checkout Kubeflow UATs
+        run: |
+          git clone https://github.com/canonical/charmed-kubeflow-uats.git ~/charmed-kubeflow-uats
+          cd ~/charmed-kubeflow-uats
+          git checkout main
+      - name: Run UATs
+        run: |
+          eval "$(pyenv init -)"
+          sg microk8s -c "tox -c ~/charmed-kubeflow-uats/ -e kubeflow -- --env params.env"
```
Comment (kimwnasptd) [inline .github/workflows/full-bundle-tests.yaml:128]: Small reminder to remove the `--env params.env` part and the code that generates this file.

It was initially used to inject ENV Vars for the Proxy in all pods generated by the UATs

File: .github/workflows/full-bundle-tests.yaml [hunk 2]
```diff
@@ -9,6 +9,18 @@ on:
       bundle-source:
```
Comment (NohaIhab) [inline .github/workflows/full-bundle-tests.yaml:9]: would it make sense to use only `bundle-source` input and figure out the `bundle-test-path` from that?
It feels redundant to do `test folder: tests-bundle/1.8` and `channel: --channel=1.8/stable`

File: .github/workflows/full-bundle-tests.yaml [hunk 3]
```diff
@@ -33,98 +60,112 @@ jobs:
 
           echo "MY_ADDONS=hostpath-storage ingress dns:$dns_server rbac registry metallb:'10.64.140.43-10.64.140.49,192.168.0.105-192.168.0.111'" >> $GITHUB_OUTPUT
 
-      - name: Install tools
+      - name: Setup aproxy
```
Comment (NohaIhab) [inline .github/workflows/full-bundle-tests.yaml:63]: Can you link here the source of these instructions in a comment for clarity?

File: tests-bundle/1.8/requirements.txt [hunk 4]
```diff
@@ -2,6 +2,6 @@ lightkube
 pytest
 pytest-operator
 kfp<2.0.0
-juju<3.0.0
+juju
```
Comment (NohaIhab) [inline tests-bundle/1.8/requirements.txt:5]: pin juju<4.0

File: tests-bundle/1.8/test_release_1-8.py [hunk 5]
```diff
@@ -46,7 +45,7 @@ async def test_deploy(ops_test: OpsTest, lightkube_client, deploy_cmd):
         'knative-serving',
         'kserve-controller',
         'kubeflow-dashboard',
-        # due to https://github.com/canonical/kubeflow-profiles-operator/issues/117
+        # due to https://github.com/c   anonical/kubeflow-profiles-operator/issues/117
```
Comment (NohaIhab) [inline tests-bundle/1.8/test_release_1-8.py:48]: nit: is this change intentional?

**Review by NohaIhab — COMMENTED** [review]
thanks @orfeas-k 
some small comments

File: .github/workflows/full-bundle-tests.yaml (hunk 3 above)
Comment (orfeas-k) [reply .github/workflows/full-bundle-tests.yaml:63]: Good catch, let me add this.

File: .github/workflows/full-bundle-tests.yaml (hunk 2 above)
Comment (NohaIhab) [reply .github/workflows/full-bundle-tests.yaml:9]: I think leveraging the script makes sense for the bundle path.
about Juju and uk8s, I can see a case where we want to specify those versions ourselves, so I think it's ok to keep them as inputs.

Comment (orfeas-k) [reply .github/workflows/full-bundle-tests.yaml:6]: Agreed. This way, other workflows that need to use this one (`workflow_call`) won't have to specify the `bundle-test-path` input too. Just pushed 2 commits for this.

File: scripts/get_release_from_bundle_source.py [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+
+    bundle_source = sys.argv[1]
+    # Bundle source input should be `--channel <channel_name>` or `--file <bundle_file>.yaml``
+    # e.g. --channel=1.8/stable or --file=releases/1.8/stable/kubeflow/bundle.yaml
+    bundle_source_starts_with_channel = re.search("^--channel", bundle_source)
+    bundle_source_starts_with_file = re.search("^--file", bundle_source)
+
+    try:
+        if bundle_source_starts_with_channel:
+            substrings = bundle_source.split('=')
```
Comment (NohaIhab) [inline scripts/get_release_from_bundle_source.py:18]: what if the input is `--channel <channel_name>` without the `=`, would this work?

**Comment by NohaIhab** [comment]
to answer my comment above, I was running the UATs from the wrong branch, because we needed the fix in https://github.com/canonical/charmed-kubeflow-uats/pull/56 
The PR is merged now so we can run the action on the main branch of the UATs.
I then tried to run the action but it was complaining about async tests, so I added the pytest async mark in a [branch](https://github.com/canonical/bundle-kubeflow/tree/test-add-asyncio-mark) to test if it is needed, and seems that it is now needed.
With the async mark, the test [failed](https://github.com/canonical/bundle-kubeflow/actions/runs/7085180179/job/19280917458) for a different reason, the deployment was not complete waiting on `argo-controller` in `maintenace` with the message `Reconciling charm: executing component kubernetes:auth-crds-cm-and-secrets`

We don't have to address these issues in this PR, but they are blocking me from testing the action. WDYT? @orfeas-k

**Comment by orfeas-k** [comment]
@NohaIhab Created the issue https://github.com/canonical/bundle-kubeflow/issues/788 documenting the flakiness of the workflow. Lmk if you think I missed something.

**Review by NohaIhab — APPROVED** [review]
lgtm, we can merge this and keep in mind to address the filed issues
